content/blog/taksim-modes-makam.md

```diff
@@ -0,0 +1,664 @@
+- Do simple mode declarations provide enough generality? (TODO) [→ modal makam, take 2](#todo)
+- Can we turn the bidirectional recipe into a Makam program, to "bidirectionalize" an existing typing declaration? (TODO) [→ writing recipes in makam](#todo)
+
+## example, naively
+
+```makam-hidden
+type_of_var : expr -> typ -> prop.
+```
+
+```makam-hidden
+%extend naive.
+type_synth : expr -> typ -> prop.
+type_check : expr -> typ -> prop.
+```
+
+As a first step, let's look at the basic bidirectional typing rules for the simply-typed lambda calculus, and try to transcribe them directly to Makam. Following the definitions in the [survey](https://arxiv.org/pdf/1908.05839) paper mentioned above (Figure 1), the rules are:
+
+<center><img src="/blog/taksim-modes-bidir-spec.svg" alt="Abstract syntax tree" width="500" /></center>
+
+Transcribing them to Makam should be simple:
+
+```makam
+type_synth X A :-
+  type_of_var X A.
+
+type_check E B :-
+  type_synth E A, unify A B.
+
+type_synth (annot E A) A :-
+  type_check E A.
+
+type_check eunit tunit.
+
+type_check (lambda E) (arrow A1 A2) :-
+  (x:expr -> type_of_var x A1 -> type_check (E x) A2).
+
+type_synth (app E1 E2) B :-
+  type_synth E1 (arrow A B),
+  type_check E1 A.
+```
+
+Here I am using the additional predicate `type_of_var` to model the context $\Gamma$.
+
+This seems to work:
+
+```makam
+type_check (lambda (fun x => x)) (arrow tunit tunit) ?
+```
+
+```makam-hidden
+>> Yes.
+```
+
+As we expect though, this does not model the bidirectional system accurately. For example,
+it is able to discover a type `T` when we are using the type-checking judgment:
+
+```makam
+type_check (lambda (fun x => x)) T ?
+```
+
+```makam-hidden
+>> Yes:
+>> T := arrow T1 T1.
+```
+
+```makam-hidden
+%end.
+```
+
+An accurate model of the bidirectional system should fail instead,
+as the type `T` is unknown and can thus not be used with the
+type-checking judgment.
+
+So instead, let's try to transcribe the system more accurately.
+
+## example, manually
+
+```makam-hidden
+%extend manual.
+type_check : expr -> typ -> prop.
+type_synth : expr -> typ -> prop.
+```
+
+To start, let us focus on just one indicative checking rule:
+
+<center><img src="/blog/taksim-modes-bidir-focused.svg" alt="Abstract syntax tree" width="300" /></center>
+
+To encode this rule properly, we need to make sure
+that the type `T` of `type_check E T` is treated as an input.
+The key difference is that
+given an unknown type like `T` as input, we should not unify it with the
+`arrow A1 A2` term; instead, we should check that `T` *already*
+looks like `arrow A1 A2`. So we should perform *pattern matching*
+rather than *unification* with `arrow A1 A2`. Fortunately,
+Makam provides us with a `pattern_match` implementation that behaves
```

Comment:
Might be handy to provide a permalink to the stdlib! Eg. https://github.com/astampoulis/makam/blob/a2815a711b0257774308cd2bae21f22ee41c3030/stdlib/reify.makam#L173-L193